File: app/sanitizers/validation/ValidationResult.java
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package sanitizers.validation;

import java.util.Collections;
import java.util.List;
import java.util.Optional;
import play.libs.Json;
import play.mvc.Result;
import play.mvc.Results;

public final class ValidationResult {

    private final List<FieldError> errors;

    private ValidationResult(List<FieldError> errors) {
        this.errors = errors;
    }

    public static ValidationResult ok() {
        return new ValidationResult(Collections.emptyList());
    }

    public static ValidationResult error(String field, String message) {
        return new ValidationResult(List.of(new FieldError(field, message)));
    }

    public static ValidationResult withErrors(List<FieldError> errors) {
        return new ValidationResult(errors);
    }

    public boolean isValid() {
        return errors.isEmpty();
    }

    public List<FieldError> getErrors() {
        return errors;
    }

    public Optional<Result> toPlayResult() {
        if (isValid()) {
            return Optional.empty();
        }
        return Optional.of(Results.badRequest(Json.toJson(errors)));
    }

    @Override
    public String toString() {
        return errors.toString();
    }
}

File: app/validation/core/ValidationResult.scala
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package validation.core

import scala.jdk.CollectionConverters.*

/** Result of validation, containing zero or more field errors.
  */
case class ValidationResult(errors: List[FieldError]):
  def isValid: Boolean            = errors.isEmpty
  def getErrors: List[FieldError] = errors
  override def toString: String   = errors.toString

object ValidationResult:
  def ok(): ValidationResult = ValidationResult(List.empty)
  def error(field: String, message: String): ValidationResult =
    ValidationResult(List(FieldError(field, message)))
  def withErrors(errors: List[FieldError]): ValidationResult =
    ValidationResult(errors)
  // Java interop: accept java.util.List
  def withErrors(errors: java.util.List[FieldError]): ValidationResult =
    ValidationResult(errors.asScala.toList)
